src/coreclr/jit/llvm.cpp

```diff
@@ -919,28 +919,40 @@ Value* Llvm::consumeValue(GenTree* node, Type* targetLlvmType)
                nodeValue->getType()->getPrimitiveSizeInBits() <= 32 && targetLlvmType->getPrimitiveSizeInBits() <= 32);
         if (nodeValue->getType()->getPrimitiveSizeInBits() < targetLlvmType->getPrimitiveSizeInBits())
         {
-            // Upcast.
-            if (varTypeIsSmall(node))
-            {
-                finalValue = varTypeIsSigned(node) ? _builder.CreateSExt(nodeValue, targetLlvmType)
-                                                   : _builder.CreateZExt(nodeValue, targetLlvmType);
-            }
-            else if (node->OperIs(GT_LCL_VAR))
-            {
-                // In IR, small locals are usually typed as INTs, while in LLVM registers we store them
-                // with their "true" type. So here we must re-extend them (as the user requests we do so).
-                LclVarDsc* varDsc = _compiler->lvaGetDesc(node->AsLclVarCommon());
-                assert(varTypeIsSmall(varDsc));
-                
-                finalValue = varTypeIsSigned(varDsc) ? _builder.CreateSExt(nodeValue, targetLlvmType)
-                                                     : _builder.CreateZExt(nodeValue, targetLlvmType);
-            }
-            else
+            var_types trueNodeType = TYP_UNDEF;
+
+            switch (node->OperGet())
             {
-                // This is the special case for relops. Ordinary codegen "just knows" they need zero-extension.
-                assert(nodeValue->getType() == Type::getInt1Ty(_llvmContext));
-                finalValue = _builder.CreateZExt(nodeValue, targetLlvmType);
+                case GT_CALL:
+                {
+                    trueNodeType = static_cast<var_types>(node->AsCall()->gtReturnType);
+                    break;
+                }
+                case GT_LCL_VAR:
+                {
+                    LclVarDsc* varDsc = _compiler->lvaGetDesc(node->AsLclVarCommon());
+                    trueNodeType = varDsc->TypeGet();
+                    break;
+                }
+                default :
+                {
+                    if (node->OperIsRelop())
+                    {
+                        // This is the special case for relops. Ordinary codegen "just knows" they need zero-extension.
+                        assert(nodeValue->getType() == Type::getInt1Ty(_llvmContext));
+                        trueNodeType = TYP_BOOL;
+                    }
+                    else
+                    {
+                        trueNodeType = node->TypeGet();
+                    }
+                }
```

Comment:
```suggestion
                case GT_CALL:
                    trueNodeType = static_cast<var_types>(node->AsCall()->gtReturnType);
                    break;

                case GT_LCL_VAR:
                    trueNodeType = _compiler->lvaGetDesc(node->AsLclVarCommon())->TypeGet();
                    break;

                case GT_EQ:
                case GT_NE:
                case GT_LT:
                case GT_LE:
                case GT_GE:
                case GT_GT:
                    // This is the special case for relops. Ordinary codegen "just knows" they need zero-extension.
                    assert(nodeValue->getType() == Type::getInt1Ty(_llvmContext));
                    trueNodeType = TYP_UBYTE;
                    break;

                default:
                    trueNodeType = node->TypeGet();
                    break;
```
(The relops can be just listed explicitly, we'll not be adding new ones any time soon)

Reply:
Thanks